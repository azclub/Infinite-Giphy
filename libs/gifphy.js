import axios from 'axios';

export const fetchGifs = async ({query = '', offset = 0, limit = 10, signal} = {}) => {
  const api_key = process.env.GIFPHY_API_KEY

  if(typeof api_key !== 'string') {
    throw new Error('ApiKeyIsRequired')
  }
  
  if(query.length === 0) {
    throw new Error('QueryParamIsRequired')
  }

  try {
    const response = await axios.get('https://api.giphy.com/v1/gifs/search',{
      params: {
        api_key,
        q: query,
        limit,
        offset,
        rating: 'G',
        lang:'en'
      },
    })
    return response.data.data
  } catch(e) {
    throw new Error(`RequestFailed: ${e.message}`)
  }
}