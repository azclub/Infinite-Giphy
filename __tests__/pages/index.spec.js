import { shallow, mount } from 'enzyme'

jest.mock('../../libs/gifphy')
import { fetchGifs } from '../../libs/gifphy'

jest.mock('lodash')
import { debounce } from 'lodash';

import renderer from 'react-test-renderer'

import App from '../../pages/index'

describe('App', () => {
  let app;
  beforeEach(() => {
    fetchGifs.mockImplementation(() => [])
    debounce.mockImplementation(cb => cb)
    jest.spyOn(window, 'addEventListener')
    app = mount(<App />)
  })

  afterEach(() => {
    jest.resetAllMocks();
  })

  it('fetches GIF when mount', () => {  
    expect(fetchGifs).toHaveBeenCalled()
  })

  it('attaches scroll event handler on window', () => {  
    expect(window.addEventListener).toHaveBeenCalledWith(
      'scroll',
      app.instance().handelInfiniteScroll,
      { passive: true }
    );
  })

  it('has a button that toggles column type', () => {
    expect(app.state().column).toBe(3)
    app.find('.giphy__column-toggle').simulate('click')
    expect(app.state().column).toBe(1)
  })

  it('has a input field that search relevant GIFs', () => {
    expect(app.state().query).toBe('pun')
    app
      .find('.filter__input')
      .simulate('change', { target: { value: 'dogs' }});
    
    expect(debounce).toHaveBeenCalled()
    expect(fetchGifs).toHaveBeenCalled()
    expect(app.state().query).toBe('dogs')
  })

  describe('Snapshot Testing', () => {
    it('renders loading state', () => {
      const component = renderer.create(<App />)
      const tree = component.toJSON()
      expect(tree).toMatchSnapshot()
    })
  })
})


describe('#handelInfiniteScroll', () => {
  const defaultWindowInnerHeight = window.innerHeight
  const defaultWindowScrollY = window.scrollY
  let documentBodyOffsetHeight = 1000

  beforeEach(() => {
    fetchGifs.mockImplementation(() => [])
    debounce.mockImplementation(cb => cb)
    jest
      .spyOn(document.body, 'offsetHeight', 'get')
      .mockImplementation(() => documentBodyOffsetHeight);
  })

  afterEach(() => {
    jest.resetAllMocks();
    window.innerHeight = defaultWindowInnerHeight
    window.scrollY = defaultWindowScrollY
  })

  it('calls fetches GIFs when scroll to bottom', () => {
    window.innerHeight = 800
    window.scrollY = 0
    shallow(<App />).instance().handelInfiniteScroll()
    // First #fetchGifs is called when mount.
    expect(fetchGifs.mock.calls.length).toBe(2)
  })

  it('will not fetches GIFs when scroll did not reachbottom', () => {
    window.innerHeight = 0
    window.scrollY = 0
    shallow(<App />).instance().handelInfiniteScroll()
    // First #fetchGifs is called when mount.
    expect(fetchGifs.mock.calls.length).toBe(1)
  })
})
