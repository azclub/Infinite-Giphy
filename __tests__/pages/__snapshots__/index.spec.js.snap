// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App Snapshot Testing renders loading state 1`] = `
<div>
  <div
    className="giphy giphy--three-column"
  >
    <div
      className="giphy__filter"
    >
      <h1
        className="filter__title"
      >
        Loading...
      </h1>
      <input
        className="filter__input"
        onChange={[Function]}
        placeholder="Search GIFs by Keywords (e.g puns, kittens, dogs, etc...)"
        tabIndex="0"
        type="text"
      />
      <button
        className="giphy__column-toggle"
        onClick={[Function]}
      >
        Toggle Column
      </button>
    </div>
    <div
      className="giphy__result--loading"
    >
      <img
        alt="Loading"
        src="/images/loading.svg"
      />
    </div>
  </div>
</div>
`;
